syntax = "proto3";

package ai.verta.modeldb.versioning;
option go_package = "github.com/VertaAI/modeldb/protos/gen/go/protos/public/modeldb/versioning";

option java_multiple_files = true;

message EnvironmentBlob {
    oneof content {
        PythonEnvironmentBlob python = 1;
        DockerEnvironmentBlob docker = 2;
    }
    repeated EnvironmentVariablesBlob environment_variables = 3;
    repeated string command_line = 4;
}

message EnvironmentVariablesBlob {
    string name = 1;
    string value = 2;
}

message VersionEnvironmentBlob {
    int32 major = 1;
    int32 minor = 2;
    int32 patch = 3;
    string suffix = 4;
}

message PythonRequirementEnvironmentBlob {
    string library = 1;
    string constraint = 2;
    VersionEnvironmentBlob version = 3;
}

message PythonEnvironmentBlob {
    VersionEnvironmentBlob version = 1;
    repeated PythonRequirementEnvironmentBlob requirements = 2;
    repeated PythonRequirementEnvironmentBlob constraints = 3;
}

message DockerEnvironmentBlob {
    string repository = 1;
    string tag = 2;
    string sha = 3;
}

message EnvironmentDiff {
    oneof content {
        PythonEnvironmentDiff python = 1;
        DockerEnvironmentDiff docker = 2;
    }
    // Should only contain environment variables unique to A/B (considering all fields)
    repeated EnvironmentVariablesBlob environment_variables_A = 3;
    repeated EnvironmentVariablesBlob environment_variables_B = 4;

    repeated string command_line_A = 5;
    repeated string command_line_B = 6;
}

message PythonEnvironmentDiff {
    // Should only contain requirements and constraints unique to A/B (considering all fields)
    PythonEnvironmentBlob A = 1;
    PythonEnvironmentBlob B = 2;
}

message DockerEnvironmentDiff {
    DockerEnvironmentBlob A = 1;
    DockerEnvironmentBlob B = 2;
}
